### PR TITLE
using regex for path split

### LINKED ISSUE #349: update_fact fails when path contains a key with a bracket in its name
##### SUMMARY

update_fact fails to update a fact where a key in the referenced path contains a bracket. In particular, I'm working with a dictionary which has a key called `[runners]`.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

update_fact

##### ANSIBLE VERSION

```paste below
ansible [core 2.14.3]
  config file = None
  configured module search path = ['/home/jarl/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = /home/jarl/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0] (/usr/bin/python3)
  jinja version = 3.1.2
  libyaml = True
```

##### COLLECTION VERSION

```paste below
# /home/jarl/.ansible/collections/ansible_collections
Collection    Version
------------- -------
ansible.utils 4.0.0  

# /usr/lib/python3/dist-packages/ansible_collections
Collection    Version
------------- -------
ansible.utils 2.9.0
```

##### CONFIGURATION

```paste below
CONFIG_FILE() = None
```

##### OS / ENVIRONMENT

Debian 12 on both controller and target, python version as above

##### STEPS TO REPRODUCE

```yaml
- hosts: 127.0.0.1
  connection: local
  tasks:
    - name: set fact
      set_fact:
        my_dict: "{{ { \"[runners]\": \"old value\" } }}"

    - name: update fact
      ansible.utils.update_fact:
        updates:
          - path: my_dict["[runners]"].value
            value: "new value"
```

##### EXPECTED RESULTS

The fact should update the given path.

##### ACTUAL RESULTS

The update fails, claiming it cannot find a key in my_dict called `[runners` (note missing closing bracket).

```paste below
ansible-playbook [core 2.14.3]
  config file = None
  configured module search path = ['/home/jarl/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = /home/jarl/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible-playbook
  python version = 3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0] (/usr/bin/python3)
  jinja version = 3.1.2
  libyaml = True
No config file found; using defaults
setting up inventory plugins
host_list declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
script declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
auto declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
yaml declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
ini declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
toml declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not
match 'all'
Loading collection ansible.utils from /home/jarl/.ansible/collections/ansible_collections/ansible/utils
Loading callback plugin default of type stdout, v2.0 from /usr/lib/python3/dist-packages/ansible/plugins/callback/default.py
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.

PLAYBOOK: playbook.yml *********************************************************************************************
Positional arguments: playbook.yml
verbosity: 4
connection: smart
timeout: 10
become_method: sudo
tags: ('all',)
inventory: ('/etc/ansible/hosts',)
forks: 5
1 plays in playbook.yml

PLAY [127.0.0.1] ***************************************************************************************************

TASK [Gathering Facts] *********************************************************************************************
task path: /home/jarl/playbook.yml:1
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: jarl
<127.0.0.1> EXEC /bin/sh -c 'echo ~jarl && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/jarl/.ansible/tmp `"&& mkdir "` echo /home/jarl/.ansible/tmp/ansible-tmp-1712307002.7893507-88412-109018197604426 `" && echo ansible-tmp-1712307002.7893507-88412-109018197604426="` echo /home/jarl/.ansible/tmp/ansible-tmp-1712307002.7893507-88412-109018197604426 `" ) && sleep 0'
Using module file /usr/lib/python3/dist-packages/ansible/modules/setup.py
<127.0.0.1> PUT /home/jarl/.ansible/tmp/ansible-local-8840847f5qwh8/tmpvorlhftj TO /home/jarl/.ansible/tmp/ansible-tmp-1712307002.7893507-88412-109018197604426/AnsiballZ_setup.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/jarl/.ansible/tmp/ansible-tmp-1712307002.7893507-88412-109018197604426/ /home/jarl/.ansible/tmp/ansible-tmp-1712307002.7893507-88412-109018197604426/AnsiballZ_setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python3 /home/jarl/.ansible/tmp/ansible-tmp-1712307002.7893507-88412-109018197604426/AnsiballZ_setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/jarl/.ansible/tmp/ansible-tmp-1712307002.7893507-88412-109018197604426/ > /dev/null 2>&1 && sleep 0'
ok: [127.0.0.1]

TASK [set fact] ****************************************************************************************************
task path: /home/jarl/playbook.yml:4
ok: [127.0.0.1] => {
    "ansible_facts": {
        "my_dict": {
            "[runners]": "old value"
        }
    },
    "changed": false
}

TASK [update fact] *************************************************************************************************
task path: /home/jarl/playbook.yml:8
The full traceback is:
Traceback (most recent call last):
  File "/home/jarl/.ansible/collections/ansible_collections/ansible/utils/plugins/action/update_fact.py", line 121, in set_value
    new_obj = obj[first]
              ~~~^^^^^^^
KeyError: '[runners'
fatal: [127.0.0.1]: FAILED! => {
    "changed": false,
    "msg": "Error: the key '[runners' was not found in {'[runners]': 'old value'}."
}

PLAY RECAP *********************************************************************************************************
127.0.0.1                  : ok=2    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   


```

### PR DESCRIPTION
**Pull request by eberna**
##### SUMMARY

Fixes #349

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

ansible.utils.update_fact
##### ADDITIONAL INFORMATION

If a key in the path contains a square bracket the module does not separate the path correctly and if fails. This request changes the parser part to use regular expressions. 

```
ansible@lnx-controller:~/ansible_collections/ansible/utils$ ansible-playbook -vv  main.yaml
ansible-playbook [core 2.16.3]
  config file = /home/ansible/ansible_collections/ansible/utils/ansible.cfg
  configured module search path = ['/home/ansible/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/ansible/.local/pipx/venvs/ansible/lib/python3.11/site-packages/ansible
  ansible collection location = /home/ansible/ansible_collections
  executable location = /home/ansible/.local/bin/ansible-playbook
  python version = 3.11.2 (main, May  2 2024, 11:59:08) [GCC 12.2.0] (/home/ansible/.local/pipx/venvs/ansible/bin/python)
  jinja version = 3.1.2
  libyaml = True
Using /home/ansible/ansible_collections/ansible/utils/ansible.cfg as config file
[DEPRECATION WARNING]: [defaults]collections_paths option, does not fit var naming standard, use the singular form collections_path instead. This feature
 will be removed from ansible-core in version 2.19. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'
running playbook inside collection ansible.utils
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.

PLAYBOOK: main.yaml **************************************************************************************************************************************
1 plays in main.yaml

PLAY [127.0.0.1] *****************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************
task path: /home/ansible/ansible_collections/ansible/utils/main.yaml:1
ok: [127.0.0.1]

TASK [set fact] ******************************************************************************************************************************************
task path: /home/ansible/ansible_collections/ansible/utils/main.yaml:4
ok: [127.0.0.1] => {"ansible_facts": {"my_dict": {"[runners]": "old value"}}, "changed": false}

TASK [update fact] ***************************************************************************************************************************************
task path: /home/ansible/ansible_collections/ansible/utils/main.yaml:8
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: KeyError: '[runners'
fatal: [127.0.0.1]: FAILED! => {"changed": false, "msg": "Error: the key '[runners' was not found in {'[runners]': 'old value'}."}

PLAY RECAP ***********************************************************************************************************************************************
127.0.0.1                  : ok=2    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

ansible@lnx-controller:~/ansible_collections/ansible/utils$ ansible-playbook -vv  main.yaml
ansible-playbook [core 2.16.3]
  config file = /home/ansible/ansible_collections/ansible/utils/ansible.cfg
  configured module search path = ['/home/ansible/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/ansible/.local/pipx/venvs/ansible/lib/python3.11/site-packages/ansible
  ansible collection location = /home/ansible/ansible_collections
  executable location = /home/ansible/.local/bin/ansible-playbook
  python version = 3.11.2 (main, May  2 2024, 11:59:08) [GCC 12.2.0] (/home/ansible/.local/pipx/venvs/ansible/bin/python)
  jinja version = 3.1.2
  libyaml = True
Using /home/ansible/ansible_collections/ansible/utils/ansible.cfg as config file
[DEPRECATION WARNING]: [defaults]collections_paths option, does not fit var naming standard, use the singular form collections_path instead. This feature
 will be removed from ansible-core in version 2.19. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'
running playbook inside collection ansible.utils
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.

PLAYBOOK: main.yaml **************************************************************************************************************************************
1 plays in main.yaml

PLAY [127.0.0.1] *****************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************
task path: /home/ansible/ansible_collections/ansible/utils/main.yaml:1
ok: [127.0.0.1]

TASK [set fact] ******************************************************************************************************************************************
task path: /home/ansible/ansible_collections/ansible/utils/main.yaml:4
ok: [127.0.0.1] => {"ansible_facts": {"my_dict": {"[runners]": "old value"}}, "changed": false}

TASK [update fact] ***************************************************************************************************************************************
task path: /home/ansible/ansible_collections/ansible/utils/main.yaml:8
changed: [127.0.0.1] => {"changed": true, "my_dict": {"[runners]": "new value"}}

TASK [debug] *********************************************************************************************************************************************
task path: /home/ansible/ansible_collections/ansible/utils/main.yaml:15
ok: [127.0.0.1] => {
    "my_dict": {
        "[runners]": "old value"
    }
}

TASK [debug] *********************************************************************************************************************************************
task path: /home/ansible/ansible_collections/ansible/utils/main.yaml:16
ok: [127.0.0.1] => {
    "updated": {
        "changed": true,
        "failed": false,
        "my_dict": {
            "[runners]": "new value"
        }
    }
}

PLAY RECAP ***********************************************************************************************************************************************
127.0.0.1                  : ok=5    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```
Hope this may help, thank you.

### CODE REVIEW

**Comment by Ruchip16** [comment]
@eberna can you please add specific tests as well to validate the change?

**Comment by eberna** [comment]
> @eberna can you please add specific tests as well to validate the change?

Integration or unit test? or both? 
thanks

**Comment by Ruchip16** [comment]
> > @eberna can you please add specific tests as well to validate the change?
> 
> Integration or unit test? or both? thanks

I am fine with both or anything that would validate this particular change.

**Comment by eberna** [comment]
> > > @eberna can you please add specific tests as well to validate the change?
> > 
> > 
> > Integration or unit test? or both? thanks
> 
> I am fine with both or anything that would validate this particular change.

done, I hope it's okay, in case let me know and I'll edit.

File: plugins/action/update_fact.py [hunk 1]
```diff
@@ -70,39 +70,27 @@ def _field_split(path):
...
-                    # make numbers numbers
-                    fields.append(ast.literal_eval(field))
-                except Exception:
-                    # or strip the quotes
-                    fields.append(re.sub("['\"]", "", field))
-        except IndexError:
-            # pop'ed past the end of the que
-            # so add the final field
+        # regex match dotted values and square brackets
+        regex = re.compile(r"^[^\[\].]+|^\[[\'\"].+?[\'\"]\]|^\[.+?\]")
```
Comment (Ruchip16) [inline plugins/action/update_fact.py:76]: @eberna any particular reason for adding regex here instead of the existing code behaviour?

**Comment by Ruchip16** [comment]
pre-commit.ci autofix

File: plugins/action/update_fact.py (hunk 1 above)
Comment (eberna) [reply plugins/action/update_fact.py:76]: If a key in the path contains a square bracket the module does not separate the path correctly and if fails. This request changes the parser part to use regular expressions.

File: plugins/action/update_fact.py [hunk 2]
```diff
@@ -71,39 +71,27 @@ def _field_split(path):
...
```
Comment (TheNova22) [inline plugins/action/update_fact.py:77]: Doesnt branch 3 (`^\[.+?\]`) and branch 2 covers similar cases? (`^\[[\'\"].+?[\'\"]\]`)

Branch 3 seems like superset of branch 2. So can branch 2 be removed in that case?
Branch 2 matches - ['key1'], ["key2"]
Branch 3 matches - [key0],['key1'], ["key2"]

File: tests/integration/targets/utils_update_fact/tasks/main.yaml [hunk 3]
```diff
@@ -6,6 +6,9 @@
         c:
           - 1
           - 2
+        d:
```
Comment (TheNova22) [inline tests/integration/targets/utils_update_fact/tasks/main.yaml:9]: Can some more complicated/edge cases be added, just to ensure no future change breaks it.
Example:
```
    - name: Set a fact with multiple bracket variations
      ansible.builtin.set_fact:
        bracket_edge_cases:
          "[[double]]": value1
          "]]right]]": value2
          "[prefix": value3
          "suffix]": value4
          "{curlies}": value 5
          "[parent]":
            - regular:
                "[bracket]": value6
    - name: update fact
      ansible.utils.update_fact:
        updates:
          - path: bracket_edge_cases["[[double]]"]
            value: "newval1"
          - path: bracket_edge_cases["]]right]]"]
            value: "newval2"
          - path: bracket_edge_cases["[prefix"]
            value: "newval3"
          - path: bracket_edge_cases["suffix]"]
            value: "newval4"
          - path: bracket_edge_cases["{curlies}"]
            value: "newval5"
          - path: bracket_edge_cases["[parent]"][0]["regular"]["[bracket]"]
            value: "newval6"
```

**Review by TheNova22 — CHANGES_REQUESTED** [review]

File: plugins/action/update_fact.py (hunk 2 above)
Comment (eberna) [reply plugins/action/update_fact.py:77]: Branch 2 covers exactly what the issues was, the presence of brackets inside the quotes like:
['tes[]t1']['test2']
['tes]t1'][test2]
['tes[t1'].test2

Since some time has passed since i wrote it, i tested it again: https://regex101.com/r/qDRrV4/1
This one does not look to me a valid list: [key0],'key1'], ["key2"], also it does not covers nested lists but as far i remember neither did before. Thanks :-)

Comment (TheNova22) [reply plugins/action/update_fact.py:77]: Oh thanks! It's a typo from my end, it should be: 
> Branch 3 matches - [key0],['key1'], ["key2"]

And that is right, nested scenario and even the base case wasn't covered before 
Thanks for clearing it for me! :) 
Please check on the other comment to add some edge cases like the one mentioned here. Marking this comment as resolved.

File: tests/integration/targets/utils_update_fact/tasks/main.yaml (hunk 3 above)
Comment (eberna) [reply tests/integration/targets/utils_update_fact/tasks/main.yaml:9]: Added more tests as requested, thank for the review!

Comment (TheNova22) [reply tests/integration/targets/utils_update_fact/tasks/main.yaml:9]: Yay, thanks for looking into it and making this change @eberna! :rocket:

**Review by TheNova22 — APPROVED** [review]

**Review by rohitthakur2590 — APPROVED** [review]

**Comment by rohitthakur2590** [comment]
@eberna  thank you for your valuable contribution.
